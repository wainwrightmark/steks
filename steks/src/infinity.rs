use itertools::Itertools;
use rand::{rngs::StdRng, SeedableRng};

use crate::prelude::*;

pub fn get_initial_shapes(seed: u64) -> Vec<ShapeIndex> {
    let mut rng: StdRng = StdRng::seed_from_u64(seed);
    let mut shapes: Vec<ShapeIndex> = vec![];
    for _ in 0..INFINITE_MODE_STARTING_SHAPES {
        shapes.push(ShapeIndex::random_no_circle(&mut rng));
    }

    shapes
}

pub fn get_all_shapes<'a>(seed: u64, total_shapes: usize) -> Vec<ShapeCreationData> {
    let mut collected = get_initial_shapes(seed);
    let mut results: Vec<ShapeCreationData> = collected.iter().map(|x|ShapeCreationData::from(*x)) .clone().collect_vec();
    collected.sort();

    while collected.len() < total_shapes {
        let mut hash: u64 = 0;
        for s in collected.iter() {
            hash |= hash.wrapping_mul(97).wrapping_add(s.0 as u64)
        }
        let next = ShapeIndex::from_seed_no_circle(hash);

        results.push(next.into());
        collected.push(next);
        collected.sort();
    }

    results
}
