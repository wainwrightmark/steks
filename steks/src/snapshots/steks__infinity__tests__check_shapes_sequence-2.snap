---
source: steks/src/infinity.rs
expression: indices
---
[
    5,
    11,
    14,
    14,
    15,
    11,
    10,
    1,
    17,
    1,
    8,
    14,
    6,
    16,
    15,
    11,
    10,
    5,
    12,
    11,
    4,
    2,
    20,
    3,
    5,
    20,
    8,
    18,
    10,
    17,
    19,
    1,
    7,
    5,
    12,
    12,
    10,
    7,
    8,
    2,
    16,
    14,
    20,
    1,
    10,
    1,
    2,
    18,
    16,
    19,
]
